src/dune_rules/opam_create.ml

```diff
@@ -57,17 +57,53 @@ let default_build_command =
   [ "dune" "install" "-p" name "--create-install-files" name ]
 ]
 |}))
+  and from_3_0 =
+    let fname = "<internal>" in
+    let parse s = lazy (Opam_file.parse_value (Lexbuf.from_string ~fname s)) in
+    let subst = parse {| [ "dune" "subst" ] {dev} |} in
+    let build =
+      parse
+        {|
+  [ "dune" "build" "-p" name "-j" jobs
+      "@install"
+      "@runtest" {with-test}
+      "@doc" {with-doc}
+  ]
+|}
+    in
+    let install =
+      parse {| [ "dune" "install" "-p" name "--create-install-files" name ] |}
+    in
+    fun ~with_sites ->
+      let dumb_pos = (fname, 0, 0) in
+      if with_sites then
+        let build =
+          match Lazy.force build with
+          | OpamParserTypes.List (op, l) ->
+            OpamParserTypes.List
+              (op, l @ [ String (op, "--promote-install-files=false") ])
+          | _ -> assert false
+        in
+        OpamParserTypes.List
+          (dumb_pos, [ Lazy.force subst; build; Lazy.force install ])
+      else
+        OpamParserTypes.List (dumb_pos, [ Lazy.force subst; Lazy.force build ])
```

Comment:
I think this whole section might be easier to understand by duplicating the lines and doing something similar to:
https://github.com/ocaml/dune/pull/4864/files#diff-1460ab4cbbed8158e17ed39673290712f22c59fd5a949c67184a75e36255a299R86

I don't think trying to remove duplication here is particularly readable or wanted.

Reply:
ok fair enough. It looks better now.